Issue899 accname from heading

jongund opened the pull request:
Update 2023-Jan-25: This PR abandoned in favor of #1860 and can be closed.

No longer attached to https://github.com/w3c/accname/issues/138 (Note: PR modified by @cookiecrook)

[Proposed definition of accessible name from heading](https://raw.githack.com/w3c/aria/issue899-accname-from-heading/index.html#namecalculation)

[Proposed list of roles that support name from heading](https://raw.githack.com/w3c/aria/issue899-accname-from-heading/index.html#namefromheading)

## Non-Landmark Roles Authoring Techniques
* [Proposed changes to the `alertdialog` role authoring requirements](https://raw.githack.com/w3c/aria/issue899-accname-from-heading/index.html#alertdialog)
* [Proposed changes to the `article` role authoring requirements](https://raw.githack.com/w3c/aria/issue899-accname-from-heading/index.html#article)
* [Proposed changes to the `dialog` role authoring requirements](https://raw.githack.com/w3c/aria/issue899-accname-from-heading/index.html#dialog)

## Landmark Roles Authoring Techniques
* [Proposed changes to the `complementary` role authoring requirements](https://raw.githack.com/w3c/aria/issue899-accname-from-heading/index.html#complementary)
* [Proposed changes to the `navigation` role authoring requirements](https://raw.githack.com/w3c/aria/issue899-accname-from-heading/index.html#navigation)
* [Proposed changes to the `region` role authoring requirements](https://raw.githack.com/w3c/aria/issue899-accname-from-heading/index.html#region)


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***

### :boom: Error: 500 Internal Server Error :boom: ###

[PR Preview](https://github.com/tobie/pr-preview#pr-preview) failed to build. _(Last tried on Jan 20, 2021, 10:59 PM UTC)_.

<details>
<summary>More</summary>


PR Preview relies on a number of web services to run. There seems to be an issue with the following one:

:rotating_light: [Spec Generator](https://www.w3.org/2015/labs/) - Spec Generator is the web service used to build specs that rely on ReSpec.

:link: [Related URL](https://labs.w3.org/spec-generator/?type=respec&url=https%3A%2F%2Frawcdn.githack.com%2Fw3c%2Faria%2F487fb4f3d2053c865ff7ccbc93a35c7547e67ce4%2Findex.html%3FisPreview%3Dtrue)

```
waiting for function failed: timeout 30000ms exceeded
```

_If you don't have enough information above to solve the error by yourself (or to understand to which web service the error is related to, if any), please [file an issue](https://github.com/tobie/pr-preview/issues/new?title=Error%20not%20surfaced%20properly&body=See%20w3c/aria%231018.)._
</details>


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/aria/pull/1018.html" title="Last updated on Jan 26, 2023, 1:20 AM UTC (51af51b)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/aria/1018/690285a...51af51b.html" title="Last updated on Jan 26, 2023, 1:20 AM UTC (51af51b)">Diff</a>

scottaohara commented on index.html:
```diff
@@ -568,7 +569,12 @@ <h4>Roles Supporting Name from Encapsulation</h4>
+			</section>
+            <section id="namefromheading">
+                <h4>Roles Supporting Name from Heading</h4>
+                <div id="index_fromheading">
```
should `region` be in this list too?

jongund replied:
@scottaohara 

It is a good idea, but I am not sure how easy it would be for browsers to  implement (e.g. making an exception for `section` element) or for authors to understand how to use (e.g. not understanding they cannot use the technique when using the `section` element.   So need more discussion on this potential addition.  If you make it to easy for authors to add region landmarks, it also may contribute to landmark pollution on pages.

scottaohara replied:
right, i wasn't advocating for making it easy to add `region` landmarks.  I want to continue to avoid that.  

what i was advocating for is if an element is specifically given a `role=region` role, then the inheriting name from heading rules would make sense.

The rules for authors providing an explicit name to `section` elements to promote it to a `role=region` would still apply / not be changed.

scottaohara replied:
resolving this and will do another review in lieu of all the changes.

LJWatson commented:
I don't understand the use case for including `main` in this list. There should only be one element with a role of `main` available at any one time, so what is the rationale for giving it an accessible name?

The same is true if the container with a role of `article` represents the main content of the page (as a blog post might for example).

The proposal here says that the first descendant element with a role of `heading` will serve as the accessible name for any of the containers included on the list. What happens when two of those containers are nested, as in the following (extremely common) pattern:
```
<main>
<article>
<h1>My post</h1>
...
</article>
</main>
```

Would both the `<main>` and `<article>` elements inherit the heading as accessible name?

jongund commented:
@LJWatson 

I think the issue of nesting of elements that can receive accessible name by heading needs discussion.  In general it seems that a heading should only provide an accessible name for one element.  So in your case only the `article` role would get the accessible name.   But there maybe cases where it could be useful, but in general I think we should be more conservative in this technique providing accessible names.

scottaohara commented:
to @LJWatson's point about `main` though, when did that get added to the list?  

It's not in the [listing of elements in the issue](https://github.com/w3c/aria/issues/899#issuecomment-487607333)?

and seems `form` was dropped?

LJWatson commented:
@jongund said:
>I think the issue of nesting of elements that can receive accessible name by heading needs discussion. In general it seems that a heading should only provide an accessible name for one element. So in your case only the article role would get the accessible name. But there maybe cases where it could be useful, but in general I think we should be more conservative in this technique providing accessible names.>

I agree about being conservative with this. Whatever route is taken, I think the spec text needs to be unambiguous on the matter though.

jongund commented:
@LJWatson @scottaohara @jnurthen @mcking65 

I added a restrictions for naming only the closest ancestor of elements that can be named with headings to address the nesting issue.   I also added `region` role as a potential role to be named with header.   This would require in HTML to use `<section role="region">`  for the heading to name a `section` element.

scottaohara commented on index.html:
```diff
@@ -6585,7 +6585,8 @@ <h5>Presentational Roles Conflict Resolution</h5>
 				<p>A perceivable <rref>section</rref> containing content that is relevant to a specific, author-specified  purpose and sufficiently important that users will likely want to be able to navigate to the section easily and to have it listed in a summary of the page. Such a page summary could be generated dynamically by a user agent or assistive technology.</p>
 				<p>Authors SHOULD limit use of the region role to sections containing content with a purpose that is not accurately described by one of the other <rref>landmark</rref> roles, such as <rref>main</rref>, <rref>complementary</rref>, or <rref>navigation</rref>.</p>
-				<p>Authors MUST give each element with role region a brief label that describes the purpose of the content in the region. Authors SHOULD reference a visible label with <pref>aria-labelledby</pref> if a visible label is present. Authors SHOULD include the label inside of a heading whenever possible. The heading MAY be an instance of the standard host language heading element or an instance of  ...
+				<p>Authors MUST give each element with role region a brief label that describes the purpose of the content in the region. Authors SHOULD use a visible heading to provide a label whenever possible.  The heading MAY be an instance of the standard host language heading element or an instance of an element with role <rref>heading</rref>.  For host languages that support elments with the default r ...
```
Regarding the "For host languages...." bit, might this be reworded as:
>An explicit `role="region"` attribute must be defined to enable the heading labeling technique.

Since `section` does *not default* to `role=region`, unless given an `aria-label` or `aria-labelledby`, which would then take precedent over this technique.

scottaohara commented:
Apologies @jongund and @mcking65, i was completely mistaken on thinking an element with `role=region` would be exposed without a name.  No screen reader / browser pairing I currently have access to exposed a nameless element with `role=region` (JAWS, NVDA, VO on macOS and iOS).  

So now, as Matt did, I'm questioning my ask to include `role=region` in this list.  I think I could go either way on it now.

jongund commented:
@cookiecrook 
I made the changes based on today's teleconference to require the heading to be the first child element and removed the `main` role from the list of roles that can be named using a 'heading` role.  But I think the restriction of first child would probably not satisfy the original purpose of the [issue 899](https://github.com/w3c/aria/issues/899) of labeling elements with the `article` role.  At the FTF this spring James Craig described the use of `article` as messy and restricting the heading to be first child would probably not help very much in giving unlabeled 'articles' accessible names.

Seems like we could use two potentially different definitions for name from heading:
1. Stricter requiring first child for the  `navigation`, `complementary` and `region` roles 
2. Looser requirement allowing any descendant heading for `article`, `dialog` and `alertdialog` roles, with a restriction of the heading apply only to only the closest ancestor.

cookiecrook commented:
> @jongund wrote
> I made the changes based on today's teleconference to require the heading to be the first child element

Option 1 is way too prescriptive and will not solve the issue for most of the suggested use cases. As one anecdotal example, Facebook articles (_if_ they use headings at all), never place them as the first child of the article. Often, even the first leaf node with content is also not a heading... It's an image, or a link, or another control.

Option 2 is a little better, but "~first heading" would be better than "~nearest in the DOM hierarchy." For example:

article 
 - div > div > heading (expected heading to choose, but excluded due to algorithm)
 - content
 - div > heading (included b/c nearer in tree, even though later in the DOM content)

cookiecrook commented:
Also, first heading may be more performant than nearest by depth, since the latter requires loading more content to calculate the comparison.

jongund commented:
@LJWatson @scottaohara @cookiecrook 

Updated name from heading to account for first child requirement for landmark roles and more general requirement for other roles.

cookiecrook commented on index.html:
```diff
@@ -6549,7 +6584,7 @@ <h5>Presentational Roles Conflict Resolution</h5>
 				<p>A perceivable <rref>section</rref> containing content that is relevant to a specific, author-specified  purpose and sufficiently important that users will likely want to be able to navigate to the section easily and to have it listed in a summary of the page. Such a page summary could be generated dynamically by a user agent or assistive technology.</p>
 				<p>Authors SHOULD limit use of the region role to sections containing content with a purpose that is not accurately described by one of the other <rref>landmark</rref> roles, such as <rref>main</rref>, <rref>complementary</rref>, or <rref>navigation</rref>.</p>
-				<p>Authors MUST give each element with role region a brief label that describes the purpose of the content in the region. Authors SHOULD reference a visible label with <pref>aria-labelledby</pref> if a visible label is present. Authors SHOULD include the label inside of a heading whenever possible. The heading MAY be an instance of the standard host language heading element or an instance of  ...
+				<p>Authors MUST give each element with role region a brief label that describes the purpose of the content in the region. Authors SHOULD use a visible heading as the first child element of the region to provide a label whenever possible.  The heading MAY be an instance of the standard host language heading element or an instance of an element with role <rref>heading</rref>.  For host language ...
```
```diff
- For host languages that support elments with the default role of region (e.g. HTML <code>section</code>), an explict region role must be defined (e.g. <code>role=<q>region</q></code>) to enable the heading labeling technique. 
```
What is this one intending to solve? This might break the original use case in #899.

cookiecrook reviewed (changes requested):
Original suggestion in #899 was to allow better flexibility for implementations to heuristically improve the experience for users. I think this PR is overly rigid, and does not solve the original intention, but it might be reconcilable with changes.

I would still prefer the flexible solution closer to what I originally proposed. As an example, ML-based heuristic discovery of label is probably going to guess better than anything we can spec around author errors like this. Over-specifying may hinder those improvements.

cookiecrook commented on index.html:
```diff
@@ -533,6 +533,7 @@ <h3>Accessible Name Calculation</h3>
 						<li>author: name comes from values provided by the author in explicit markup features such as the <pref>aria-label</pref> attribute, the <pref>aria-labelledby</pref> attribute, or the host language labeling mechanism, such as the <code>alt</code> or <code>title</code> attributes in <abbr title="Hypertext Markup Language">HTML</abbr>, with HTML <code>title</code> attribute having the lowest  ...
 						<li>contents: name comes from the text value of the <a>element</a> node. Although this may be allowed in addition to "author" in some <a>roles</a>, this is used in content only if higher priority "author" features are not provided. Priority is defined by the <a href="https://www.w3.org/TR/accname-aam-1.1/#mapping_additional_nd_te">accessible name and description computation</a> algorithm.   ...
             <li>encapsulation: name comes from the text value of the <a>element</a> node with role <code>label</code> that is the closest ancestor. Although "encapsulation" may be allowed in addition to "author" and "contents" in some <a>roles</a>, "encapsulation" is used only if higher priority "author" features are not provided. Priority is defined by the <a href="https://www.w3.org/TR/accname- ...
+            <li>heading: name comes from the text value of the first descendant (i.e. depth first) <a>element</a> node with the role of <code>heading</code>, with landmark roles heading elements used for naming are restricted to first child node.  A heading can only be used to name its nearest ancestor element that allows name from heading.  Although "heading" may be allowed in addition to "autho ...
```
WG consensus remove:

```diff
- with landmark roles heading elements used for naming are restricted to first child node.
```

cookiecrook commented on index.html:
```diff
@@ -2752,7 +2776,12 @@ <h2>Definition of Roles</h2>
-						<td class="role-namefrom">author</td>
+						<td class="role-namefrom">
+							<ul>
+								<li>heading</li>
+								<li>author</li>
```
```suggestion
								<li>author</li>
								<li>heading</li>
```

cookiecrook commented on index.html:
```diff
@@ -533,6 +533,7 @@ <h3>Accessible Name Calculation</h3>
 						<li>author: name comes from values provided by the author in explicit markup features such as the <pref>aria-label</pref> attribute, the <pref>aria-labelledby</pref> attribute, or the host language labeling mechanism, such as the <code>alt</code> or <code>title</code> attributes in <abbr title="Hypertext Markup Language">HTML</abbr>, with HTML <code>title</code> attribute having the lowest  ...
 						<li>contents: name comes from the text value of the <a>element</a> node. Although this may be allowed in addition to "author" in some <a>roles</a>, this is used in content only if higher priority "author" features are not provided. Priority is defined by the <a href="https://www.w3.org/TR/accname-aam-1.1/#mapping_additional_nd_te">accessible name and description computation</a> algorithm.   ...
             <li>encapsulation: name comes from the text value of the <a>element</a> node with role <code>label</code> that is the closest ancestor. Although "encapsulation" may be allowed in addition to "author" and "contents" in some <a>roles</a>, "encapsulation" is used only if higher priority "author" features are not provided. Priority is defined by the <a href="https://www.w3.org/TR/accname- ...
+            <li>heading: name comes from the text value of the first descendant (i.e. depth first) <a>element</a> node with the role of <code>heading</code>, with landmark roles heading elements used for naming are restricted to first child node.  A heading can only be used to name its nearest ancestor element that allows name from heading.  Although "heading" may be allowed in addition to "autho ...
```
```suggestion
            <li>heading: name comes from the text value of the first descendant (i.e. depth first) <a>element</a> node with the role of <code>heading</code>. Although "heading" may be allowed in addition to "author" in some <a>roles</a>, "heading" is used in content only if higher priority "author" features are not provided. Priority is defined by the <a href="https://www.w3.org/TR/accname-aam-1.1/#mapping_additional_nd_te">accessible name and description computation</a> algorithm.  [[ACCNAME-1.1]]</li>
```

cookiecrook commented on index.html:
```diff
@@ -915,7 +922,12 @@ <h2>Definition of Roles</h2>
-						<td class="role-namefrom">author</td>
+						<td class="role-namefrom">
+							<ul>
+								<li>heading</li>
+								<li>author</li>
```
```suggestion
								<li>author</li>
								<li>heading</li>
```

cookiecrook commented on index.html:
```diff
@@ -5644,7 +5674,12 @@ <h4>Plain HTML or Polyfill DOM Result of the MathML Quadratic Formula</h4>
-						<td class="role-namefrom">author</td>
+						<td class="role-namefrom">
+							<ul>
+								<li>heading</li>
+								<li>author</li>
```
```suggestion
								<li>author</li>
								<li>heading</li>
```

cookiecrook commented on index.html:
```diff
@@ -6605,7 +6640,12 @@ <h5>Presentational Roles Conflict Resolution</h5>
-						<td class="role-namefrom">author</td>
+						<td class="role-namefrom">
+							<ul>
+								<li>heading</li>
+								<li>author</li>
```
```suggestion
								<li>author</li>
								<li>heading</li>
```

cookiecrook commented on common/script/aria.js:
```diff
@@ -351,19 +352,22 @@ require(["core/pubsubhub"], function( respecEvents ) {
                             if (node.textContent.indexOf("author") !== -1) {
                                 fromAuthor += "<li><a href=\"#" + pnID + "\" class=\"role-reference\"><code>" + content + "</code></a>" + req + "</li>";
-                            } 
+                            }
```
```suggestion
                            } 
```

cookiecrook commented on common/script/aria.js:
```diff
@@ -125,7 +125,7 @@ require(["core/pubsubhub"], function( respecEvents ) {
                             ;
                     }
-                    
+
```
```suggestion
                    
```

cookiecrook commented on common/script/aria.js:
```diff
@@ -173,11 +173,11 @@ require(["core/pubsubhub"], function( respecEvents ) {
                     if ((abstract.textContent || abstract.innerText) === "All elements of the base markup") {
                         globalSP.push({ is: type, title: title, name: content, desc: desc, prohibited: false });
-                    } 
+                    }
```
```suggestion
                    } 
```

cookiecrook commented on common/script/aria.js:
```diff
@@ -173,11 +173,11 @@ require(["core/pubsubhub"], function( respecEvents ) {
                         globalSP.push({ is: type, title: title, name: content, desc: desc, prohibited: true });
-                    } 
-                    
+                    }
+
```
```suggestion
                    } 
                    
```

cookiecrook commented on common/script/aria.js:
```diff
@@ -192,7 +192,7 @@ require(["core/pubsubhub"], function( respecEvents ) {
                     }
                 });
-                
+
```
```suggestion
                    
```

cookiecrook commented on common/script/aria.js:
```diff
@@ -317,7 +318,7 @@ require(["core/pubsubhub"], function( respecEvents ) {
                         });
                     }
-                    // are there supported states / properties in this role?  
+                    // are there supported states / properties in this role?
```
```suggestion
                    // are there supported states / properties in this role?  
```

cookiecrook commented on common/script/aria.js:
```diff
@@ -330,8 +331,8 @@ require(["core/pubsubhub"], function( respecEvents ) {
                                 var dis = $(node).hasClass("role-disallowed");
-                                attrs.push( { is: type, name: name, required: req, disallowed: dis } );                                
-                                
+                                attrs.push( { is: type, name: name, required: req, disallowed: dis } );
+
```
```suggestion
                                attrs.push( { is: type, name: name, required: req, disallowed: dis } );                                
                                
```

cookiecrook commented on common/script/aria.js:
```diff
@@ -359,11 +360,14 @@ require(["core/pubsubhub"], function( respecEvents ) {
                             }
                             if (node.textContent.indexOf("encapsulation") !== -1) {
-                                fromEncapsulation += "<li><a href=\"#" + pnID + "\" class=\"role-reference\"><code>" + content + "</code></a>" + req + "</li>"; 
+                                fromEncapsulation += "<li><a href=\"#" + pnID + "\" class=\"role-reference\"><code>" + content + "</code></a>" + req + "</li>";
```
```suggestion
                                fromEncapsulation += "<li><a href=\"#" + pnID + "\" class=\"role-reference\"><code>" + content + "</code></a>" + req + "</li>"; 
```

cookiecrook commented on common/script/aria.js:
```diff
@@ -396,7 +400,7 @@ require(["core/pubsubhub"], function( respecEvents ) {
                     }
                 };
-                    
+
```
```suggestion
                    
```

cookiecrook commented on common/script/aria.js:
```diff
@@ -445,7 +449,7 @@ require(["core/pubsubhub"], function( respecEvents ) {
                         }
                     });
-                    
+
```
```suggestion
                    
```

cookiecrook commented on common/script/aria.js:
```diff
@@ -464,7 +468,7 @@ require(["core/pubsubhub"], function( respecEvents ) {
                         }
                     };
-                        
+
```
```suggestion
                        
```

cookiecrook commented on common/script/aria.js:
```diff
@@ -511,7 +515,7 @@ require(["core/pubsubhub"], function( respecEvents ) {
                         }
                     });
-                    
+
```
```suggestion
                    
```

cookiecrook commented on common/script/aria.js:
```diff
@@ -593,15 +607,15 @@ require(["core/pubsubhub"], function( respecEvents ) {
 
                 // prune out unused rows throughout the document
-                
+
```
```suggestion
                
```

cookiecrook commented on common/script/aria.js:
```diff
@@ -593,15 +607,15 @@ require(["core/pubsubhub"], function( respecEvents ) {
-                    } else if (content === "Placeholder" 
-                               && !skipIndex 
-                               && (item.className === "role-inherited" 
+                    } else if (content === "Placeholder"
+                               && !skipIndex
+                               && (item.className === "role-inherited"
```
```suggestion
                    } else if (content === "Placeholder" 
                               && !skipIndex 
                               && (item.className === "role-inherited" 
```

cookiecrook commented on index.html:
```diff
@@ -494,7 +494,7 @@ <h3>Inherited States and Properties</h3>
 				<p>List of states and properties that are prohibited on a <a>role</a>. Authors MUST NOT specify a prohibited state or property.</p>
 				<p class="note">A host language attribute with the appropriate <a href="#implicit_semantics">implicit WAI-ARIA semantic</a> would also prohibit a state or property in this section. </p>
-			</section>		
+			</section>
```
```suggestion
			</section>		
```

cookiecrook commented on index.html:
```diff
@@ -564,11 +565,16 @@ <h4>Roles Supporting Name from Encapsulation</h4>
                 <div id="index_fromlegend">
                 </div>
-			</section>						
+			</section>
```
```suggestion
			</section>						
```

cookiecrook commented on index.html:
```diff
@@ -1518,7 +1536,7 @@ <h2>Definition of Roles</h2>
 						<th class="role-inherited-head" scope="row">Inherited States and Properties:</th>
 						<td class="role-inherited">Placeholder</td>
-					</tr>				
+					</tr>
```
```suggestion
					</tr>				
```

cookiecrook commented on index.html:
```diff
@@ -1694,9 +1712,9 @@ <h2>Definition of Roles</h2>
 								<li><pref>aria-label</pref></li>
 								<li><pref>aria-labelledby</pref></li>
-							</ul>					
+							</ul>
```
```suggestion
							</ul>					
```

cookiecrook commented on index.html:
```diff
@@ -1694,9 +1712,9 @@ <h2>Definition of Roles</h2>
+							</ul>
 						</td>
-					</tr>		
+					</tr>
```
```suggestion
					</tr>		
```

cookiecrook commented on index.html:
```diff
@@ -1980,9 +1998,9 @@ <h2>Definition of Roles</h2>
 								<li><pref>aria-label</pref></li>
 								<li><pref>aria-labelledby</pref></li>
-							</ul>					
+							</ul>
```
```suggestion
							</ul>					
```

cookiecrook commented on index.html:
```diff
@@ -1980,9 +1998,9 @@ <h2>Definition of Roles</h2>
+							</ul>
 						</td>
-					</tr>						
+					</tr>
```
```suggestion
					</tr>						
```

cookiecrook commented on index.html:
```diff
@@ -2666,9 +2690,9 @@ <h2>Definition of Roles</h2>
 								<li><pref>aria-label</pref></li>
 								<li><pref>aria-labelledby</pref></li>
-							</ul>					
+							</ul>
```
```suggestion
							</ul>					
```

cookiecrook commented on index.html:
```diff
@@ -2666,9 +2690,9 @@ <h2>Definition of Roles</h2>
+							</ul>
 						</td>
-					</tr>							
+					</tr>
```
```suggestion
					</tr>							
```

cookiecrook commented on index.html:
```diff
@@ -2992,9 +3020,9 @@ <h2>Definition of Roles</h2>
 								<li><pref>aria-label</pref></li>
 								<li><pref>aria-labelledby</pref></li>
-							</ul>					
+							</ul>
```
```suggestion
							</ul>					
```

cookiecrook commented on index.html:
```diff
@@ -2992,9 +3020,9 @@ <h2>Definition of Roles</h2>
+							</ul>
 						</td>
-					</tr>						
+					</tr>
```
```suggestion
					</tr>						
```

cookiecrook commented on index.html:
```diff
@@ -3329,9 +3357,9 @@ <h2>Definition of Roles</h2>
 								<li><pref>aria-label</pref></li>
 								<li><pref>aria-labelledby</pref></li>
-							</ul>					
+							</ul>
```
```suggestion
							</ul>					
```

cookiecrook commented on index.html:
```diff
@@ -3329,9 +3357,9 @@ <h2>Definition of Roles</h2>
+							</ul>
 						</td>
-					</tr>						
+					</tr>
```
```suggestion
					</tr>						
```

cookiecrook commented on index.html:
```diff
@@ -3930,7 +3958,7 @@ <h2>Definition of Roles</h2>
 						<th class="role-properties-head" scope="row">Supported States and Properties:</th>
 						<td class="role-properties"> </td>
-					</tr>					
+					</tr>
```
```suggestion
					</tr>					
```

cookiecrook commented on index.html:
```diff
@@ -3941,9 +3969,9 @@ <h2>Definition of Roles</h2>
 								<li><pref>aria-label</pref></li>
 								<li><pref>aria-labelledby</pref></li>
-							</ul>					
+							</ul>
```
```suggestion
							</ul>					
```

cookiecrook commented on index.html:
```diff
@@ -3941,9 +3969,9 @@ <h2>Definition of Roles</h2>
+							</ul>
 						</td>
-					</tr>						
+					</tr>
```
```suggestion
					</tr>						
```

cookiecrook commented on index.html:
```diff
@@ -6011,9 +6045,9 @@ <h5>Note regarding the ARIA 1.1 <code>none</code> role.</h5>
 								<li><pref>aria-label</pref></li>
 								<li><pref>aria-labelledby</pref></li>
-							</ul>					
+							</ul>
```
```suggestion
							</ul>					
```

cookiecrook commented on index.html:
```diff
@@ -6011,9 +6045,9 @@ <h5>Note regarding the ARIA 1.1 <code>none</code> role.</h5>
+							</ul>
 						</td>
-					</tr>					
+					</tr>
```
```suggestion
					</tr>					
```

cookiecrook commented on index.html:
```diff
@@ -6171,9 +6205,9 @@ <h5>Presentational Roles Conflict Resolution</h5>
 								<li><pref>aria-label</pref></li>
 								<li><pref>aria-labelledby</pref></li>
-							</ul>					
+							</ul>
```
```suggestion
							</ul>					
```

cookiecrook commented on index.html:
```diff
@@ -6171,9 +6205,9 @@ <h5>Presentational Roles Conflict Resolution</h5>
+							</ul>
 						</td>
-					</tr>							
+					</tr>
```
```suggestion
					</tr>							
```

cookiecrook commented on index.html:
```diff
@@ -7963,9 +8002,9 @@ <h5>Presentational Roles Conflict Resolution</h5>
 								<li><pref>aria-label</pref></li>
 								<li><pref>aria-labelledby</pref></li>
-							</ul>					
+							</ul>
```
```suggestion
							</ul>					
```

cookiecrook commented on index.html:
```diff
@@ -7963,9 +8002,9 @@ <h5>Presentational Roles Conflict Resolution</h5>
+							</ul>
 						</td>
-					</tr>	
+					</tr>
```
```suggestion
					</tr>	
```

cookiecrook commented on index.html:
```diff
@@ -8125,9 +8164,9 @@ <h5>Presentational Roles Conflict Resolution</h5>
 								<li><pref>aria-label</pref></li>
 								<li><pref>aria-labelledby</pref></li>
-							</ul>					
+							</ul>
```
```suggestion
							</ul>					
```

cookiecrook commented on index.html:
```diff
@@ -8125,9 +8164,9 @@ <h5>Presentational Roles Conflict Resolution</h5>
+							</ul>
 						</td>
-					</tr>						
+					</tr>
```
```suggestion
					</tr>						
```

cookiecrook commented on index.html:
```diff
@@ -8212,9 +8251,9 @@ <h5>Presentational Roles Conflict Resolution</h5>
 								<li><pref>aria-label</pref></li>
 								<li><pref>aria-labelledby</pref></li>
-							</ul>					
+							</ul>
```
```suggestion
							</ul>					
```

cookiecrook commented on index.html:
```diff
@@ -8212,9 +8251,9 @@ <h5>Presentational Roles Conflict Resolution</h5>
+							</ul>
 						</td>
-					</tr>						
+					</tr>
```
```suggestion
					</tr>						
```

cookiecrook commented on index.html:
```diff
@@ -12839,7 +12878,7 @@ <h3>States and Properties</h3>
 				</li>
 				<li>Otherwise, ignore the value and treat the property as not present.</li>
-			</ul>	
+			</ul>
```
```suggestion
			</ul>	
```

cookiecrook commented on common/script/aria.js:
```diff
@@ -192,7 +192,7 @@ require(["core/pubsubhub"], function( respecEvents ) {
                     }
                 });
-                
+                    
```
```suggestion
                
```

cookiecrook commented:
Pulled out the whitespace diffs and other items WG agreed on today. Still needs further review of some content and merge conflict resolution.

cookiecrook commented on index.html:
```diff
@@ -5591,6 +5619,7 @@ <h4>Plain HTML or Polyfill DOM Result of the MathML Quadratic Formula</h4>
 			<rdef>navigation</rdef>
 			<div class="role-description">
 				<p>A collection of navigational <a>elements</a> (usually links) for navigating the document or related documents.</p>
+				<p>If authors provide more than one element with the role navigation on a page, authors SHOULD give each navigation region a brief label that describes the purpose of the content in the region. Authors SHOULD use a visible or offscreen heading as the first child element of the region to provide a label whenever possible.  The heading MAY be an instance of the standard host language heading el ...
```
```suggestion
```

cookiecrook commented on index.html:
```diff
@@ -6549,7 +6583,7 @@ <h5>Presentational Roles Conflict Resolution</h5>
 				<p>A perceivable <rref>section</rref> containing content that is relevant to a specific, author-specified  purpose and sufficiently important that users will likely want to be able to navigate to the section easily and to have it listed in a summary of the page. Such a page summary could be generated dynamically by a user agent or assistive technology.</p>
 				<p>Authors SHOULD limit use of the region role to sections containing content with a purpose that is not accurately described by one of the other <rref>landmark</rref> roles, such as <rref>main</rref>, <rref>complementary</rref>, or <rref>navigation</rref>.</p>
-				<p>Authors MUST give each element with role region a brief label that describes the purpose of the content in the region. Authors SHOULD reference a visible label with <pref>aria-labelledby</pref> if a visible label is present. Authors SHOULD include the label inside of a heading whenever possible. The heading MAY be an instance of the standard host language heading element or an instance of  ...
+				<p>Authors MUST give each element with role region a brief label that describes the purpose of the content in the region. Authors SHOULD use a visible heading as the first child element of the region to provide a label whenever possible.  The heading MAY be an instance of the standard host language heading element or an instance of an element with role <rref>heading</rref>.  For host language ...
```
```suggestion
				<p>Authors MUST give each element with role region a brief label that describes the purpose of the content in the region. For host languages that support elments with the default role of region (e.g. HTML <code>section</code>), an explict region role must be defined (e.g. <code>role="region"</code>) to enable the nameFrom: heading technique.</p>
```

cookiecrook commented on index.html:
```diff
@@ -857,6 +863,7 @@ <h2>Definition of Roles</h2>
 				<p>A type of dialog that contains an alert message, where initial focus goes to an <a>element</a> within the dialog. See related <rref>alert</rref> and <rref>dialog</rref>.</p>
 				<p>Alert dialogs are used to convey messages to alert the user. The <code>alertdialog</code> <a>role</a> goes on the node containing both the alert message and the rest of the dialog. Content authors SHOULD make alert dialogs modal by ensuring that, while the <code>alertdialog</code> is shown, keyboard and mouse interactions only operate within the dialog. See <pref>aria-modal</pref>.</p>
 				<p>Unlike <rref>alert</rref>, <code>alertdialog</code> can receive a response from the user. For example, to confirm that the user understands the alert being generated. When the alert dialog is displayed, authors SHOULD set focus to an active element within the alert dialog, such as a form edit field or an OK button. The <a>user agent</a> SHOULD fire a system alert <a>event</a> through the a ...
+				<p>Authors MUST give each element with role <rref>alertdialog</rref> a brief label that describes the purpose of the content in the alert dialog. Authors SHOULD use the first visible element of the dialog with an explicit or implicit role of <rref>heading</rref> to provide a label. If a visible label is present, but not contained in the first element with a role of <rref>heading</rref>, autho ...
```
```suggestion
```

cookiecrook replied:
Missed one

cookiecrook commented on index.html:
```diff
@@ -2308,6 +2325,7 @@ <h2>Definition of Roles</h2>
 			<div class="role-description">
 				<p>A supporting section of the document, designed to be complementary to the main content at a similar level in the DOM hierarchy, but remains meaningful when separated from the main content.</p>
 				<p>There are various types of content that would appropriately have this <a>role</a>. For example, in the case of a portal, this may include but not be limited to show times, current weather, related articles, or stocks to watch. The complementary role indicates that contained content is relevant to the main content. If the complementary content is completely separable from the main content,  ...
+				<p>If authors provide more than one element with the role complementary on a page, authors SHOULD give each complementary region a brief label that describes the purpose of the content in the region. Authors SHOULD use a visible heading as the first child element of the region to provide a label whenever possible.  The heading MAY be an instance of the standard host language heading element o ...
```
```suggestion
```

cookiecrook commented on index.html:
```diff
@@ -533,6 +533,7 @@ <h3>Accessible Name Calculation</h3>
 						<li>author: name comes from values provided by the author in explicit markup features such as the <pref>aria-label</pref> attribute, the <pref>aria-labelledby</pref> attribute, or the host language labeling mechanism, such as the <code>alt</code> or <code>title</code> attributes in <abbr title="Hypertext Markup Language">HTML</abbr>, with HTML <code>title</code> attribute having the lowest  ...
 						<li>contents: name comes from the text value of the <a>element</a> node. Although this may be allowed in addition to "author" in some <a>roles</a>, this is used in content only if higher priority "author" features are not provided. Priority is defined by the <a href="https://www.w3.org/TR/accname-aam-1.1/#mapping_additional_nd_te">accessible name and description computation</a> algorithm.   ...
             <li>encapsulation: name comes from the text value of the <a>element</a> node with role <code>label</code> that is the closest ancestor. Although "encapsulation" may be allowed in addition to "author" and "contents" in some <a>roles</a>, "encapsulation" is used only if higher priority "author" features are not provided. Priority is defined by the <a href="https://www.w3.org/TR/accname- ...
+            <li>heading: name comes from the text value of the first descendant (i.e. depth first) <a>element</a> node with the role of <code>heading</code>. Although "heading" may be allowed in addition to "author" in some <a>roles</a>, "heading" is used in content only if higher priority "author" features are not provided. Priority is defined by the <a href="https://www.w3.org/TR/accname-aam-1. ...
```
```suggestion
            <li>heading: name comes from the text value of the first descendant (i.e. depth first) <a>element</a> node with the role of <code>heading</code>. Although "heading" may be allowed in addition to "author" in some <a>roles</a>, "heading" is used in content only if higher priority "author" features are not provided. Priority is defined by the <a href="https://www.w3.org/TR/accname-aam-1.1/#mapping_additional_nd_te">accessible name and description computation</a> algorithm. [[ACCNAME-1.1]]</li>
```

cookiecrook commented:
@scottaohara @accdc This may be ready for re-review now.

jnurthen commented:
> @scottaohara @accdc This may be ready for re-review now.

I think it needs all the encapsulation and legend stuff pulling out of it.

scottaohara commented on index.html:
```diff
@@ -533,6 +533,7 @@ <h3>Accessible Name Calculation</h3>
 						<li>author: name comes from values provided by the author in explicit markup features such as the <pref>aria-label</pref> attribute, the <pref>aria-labelledby</pref> attribute, or the host language labeling mechanism, such as the <code>alt</code> or <code>title</code> attributes in <abbr title="Hypertext Markup Language">HTML</abbr>, with HTML <code>title</code> attribute having the lowest  ...
 						<li>contents: name comes from the text value of the <a>element</a> node. Although this may be allowed in addition to "author" in some <a>roles</a>, this is used in content only if higher priority "author" features are not provided. Priority is defined by the <a href="https://www.w3.org/TR/accname-aam-1.1/#mapping_additional_nd_te">accessible name and description computation</a> algorithm.   ...
             <li>encapsulation: name comes from the text value of the <a>element</a> node with role <code>label</code> that is the closest ancestor. Although "encapsulation" may be allowed in addition to "author" and "contents" in some <a>roles</a>, "encapsulation" is used only if higher priority "author" features are not provided. Priority is defined by the <a href="https://www.w3.org/TR/accname- ...
```
Per @jnurthen's comment, this should be removed because this PR is only about name from heading now?

cookiecrook replied:
```suggestion
```

scottaohara commented on index.html:
```diff
@@ -533,6 +533,7 @@ <h3>Accessible Name Calculation</h3>
 						<li>contents: name comes from the text value of the <a>element</a> node. Although this may be allowed in addition to "author" in some <a>roles</a>, this is used in content only if higher priority "author" features are not provided. Priority is defined by the <a href="https://www.w3.org/TR/accname-aam-1.1/#mapping_additional_nd_te">accessible name and description computation</a> algorithm.   ...
             <li>encapsulation: name comes from the text value of the <a>element</a> node with role <code>label</code> that is the closest ancestor. Although "encapsulation" may be allowed in addition to "author" and "contents" in some <a>roles</a>, "encapsulation" is used only if higher priority "author" features are not provided. Priority is defined by the <a href="https://www.w3.org/TR/accname- ...
+            <li>heading: name comes from the text value of the first descendant (i.e. depth first) <a>element</a> node with the role of <code>heading</code>. Although "heading" may be allowed in addition to "author" in some <a>roles</a>, "heading" is used in content only if higher priority "author" features are not provided. Priority is defined by the <a href="https://www.w3.org/TR/accname-aam-1. ...
             <li>legend: name comes from the text value of the first descendant <a>element</a> node with the role of <code>legend</code>.  Although "legend" may be allowed in addition to "author" in some <a>roles</a>, "legend" is used in content only if higher priority "author" features are not provided. Priority is defined by the <a href="https://www.w3.org/TR/accname-aam-1.1/#mapping_additional_ ...
```
this one too

cookiecrook replied:
```suggestion
```

scottaohara commented on index.html:
```diff
@@ -6549,7 +6579,7 @@ <h5>Presentational Roles Conflict Resolution</h5>
 				<p>A perceivable <rref>section</rref> containing content that is relevant to a specific, author-specified  purpose and sufficiently important that users will likely want to be able to navigate to the section easily and to have it listed in a summary of the page. Such a page summary could be generated dynamically by a user agent or assistive technology.</p>
 				<p>Authors SHOULD limit use of the region role to sections containing content with a purpose that is not accurately described by one of the other <rref>landmark</rref> roles, such as <rref>main</rref>, <rref>complementary</rref>, or <rref>navigation</rref>.</p>
-				<p>Authors MUST give each element with role region a brief label that describes the purpose of the content in the region. Authors SHOULD reference a visible label with <pref>aria-labelledby</pref> if a visible label is present. Authors SHOULD include the label inside of a heading whenever possible. The heading MAY be an instance of the standard host language heading element or an instance of  ...
+				<p>Authors MUST give each element with role region a brief label that describes the purpose of the content in the region. For host languages that support elments with the default role of region (e.g. HTML <code>section</code>), an explict region role must be defined (e.g. <code>role="region"</code>) to enable the nameFrom: heading technique.</p>
```
couple typo fixes and rewording suggestion to remove the need for two e.g.,'s in the sentence, and to swap out the instance of the lowercase 'must'.
```suggestion
				<p>Authors MUST give each element with a region role a brief label that describes the purpose of the region. For host languages that support elements with the default role of region (e.g., HTML <code>section</code>), authors would need to specify an explicit <code>role="region"</code> on the element to enable the nameFrom: heading technique.</p>
```

scottaohara replied:
A similar rule should be added to `role=complementary` per the HTML AAM change to `<aside>` where it should only map to `role=complementary` if it is scoped to the `<body>` or `<main>` element.

scottaohara commented on index.html:
```diff
@@ -533,6 +533,7 @@ <h3>Accessible Name Calculation</h3>
 						<li>author: name comes from values provided by the author in explicit markup features such as the <pref>aria-label</pref> attribute, the <pref>aria-labelledby</pref> attribute, or the host language labeling mechanism, such as the <code>alt</code> or <code>title</code> attributes in <abbr title="Hypertext Markup Language">HTML</abbr>, with HTML <code>title</code> attribute having the lowest  ...
 						<li>contents: name comes from the text value of the <a>element</a> node. Although this may be allowed in addition to "author" in some <a>roles</a>, this is used in content only if higher priority "author" features are not provided. Priority is defined by the <a href="https://www.w3.org/TR/accname-aam-1.1/#mapping_additional_nd_te">accessible name and description computation</a> algorithm.   ...
             <li>encapsulation: name comes from the text value of the <a>element</a> node with role <code>label</code> that is the closest ancestor. Although "encapsulation" may be allowed in addition to "author" and "contents" in some <a>roles</a>, "encapsulation" is used only if higher priority "author" features are not provided. Priority is defined by the <a href="https://www.w3.org/TR/accname- ...
+            <li>heading: name comes from the text value of the first descendant (i.e. depth first) <a>element</a> node with the role of <code>heading</code>. Although "heading" may be allowed in addition to "author" in some <a>roles</a>, "heading" is used in content only if higher priority "author" features are not provided. Priority is defined by the <a href="https://www.w3.org/TR/accname-aam-1. ...
```
i know this may be considered a stylistic suggestion, but the preview spec presently has more instances of "i.e.," than "i.e. "
```suggestion
            <li>heading: name comes from the text value of the first descendant (i.e., depth first) <a>element</a> node with the role of <code>heading</code>. Although "heading" may be allowed in addition to "author" in some <a>roles</a>, "heading" is used in content only if higher priority "author" features are not provided. Priority is defined by the <a href="https://www.w3.org/TR/accname-aam-1.1/#mapping_additional_nd_te">accessible name and description computation</a> algorithm. [[ACCNAME-1.1]]</li>
```

scottaohara commented:
@jnurthen @cookiecrook - I'm afraid GitHub is all wonky on this PR.  There are a bunch of resolved comments with changes that @cookiecrook made that aren't showing up in the file diff for me, nor do they appear in the preview diff of the spec.  

As mentioned, I did my first pass based on what was available in the file diff this morning, but I think I need to come back to this later.   I kinda wonder if a fresh PR is in order here detangle this and help mitigate the merge conflict of the aria.js file?

accdc commented:
I'm thoroughly confused by the code spaghetti and can't pick it apart either. If someone could please condense and resubmit this I would appreciate it.

cookiecrook commented:
I have taken assignment and am working through that as time allows. I will consider the fresh PR suggestion.

cookiecrook commented:
@scottaohara wrote:

> There are a bunch of resolved comments with changes that @cookiecrook made that aren't showing up in the file diff for me, 

That might be accurate. A lot of what I did was back out changes that were unnecessary, including all the whitespace diffs and redundant guidance/requirements that was already covered elsewhere. If you have a specific change that is still in the branch but not showing up in the diff, please let me know. That'd push me over to start a new PR entirely. Note: I've turned this back into a Draft PR until I resolve the merge conflicts.

jnurthen commented:
I tried rebasing this onto main but with the myriad of commits (especially stuff that was later reverted) I’m not sure that is a good idea. IMO closing this and creating a new PR would be the best course of action.

cookiecrook commented:
This PR is abandoned in favor of #1860 and can be closed.
